src/hotspot/share/opto/rangeinference.hpp

```diff
@@ -0,0 +1,185 @@
+// here may be arbitrary and need to be canonicalized to construct a
+// TypeInt/TypeLong
+template <class S, class U>
+class TypeIntPrototype {
+public:
+  static_assert(std::is_signed<S>::value, "");
+  static_assert(std::is_unsigned<U>::value, "");
+  static_assert(sizeof(S) == sizeof(U), "");
+
+  RangeInt<S> _srange;
+  RangeInt<U> _urange;
+  KnownBits<U> _bits;
+
+private:
+  friend class TypeInt;
+  friend class TypeLong;
+
+  template <class T1, class T2>
+  friend void test_canonicalize_constraints_simple();
+
+  template <class T1, class T2>
+  friend void test_canonicalize_constraints_random();
+
+  // A canonicalized version of a TypeIntPrototype, if the prototype represents
+  // an empty type, _present is false, otherwise, _data is canonical
+  class CanonicalizedTypeIntPrototype {
+  public:
+    bool _present; // whether this is an empty set
+    TypeIntPrototype<S, U> _data;
+
+    static CanonicalizedTypeIntPrototype make_empty() {
+      return {false, {}};
+    }
+  };
+
+  CanonicalizedTypeIntPrototype canonicalize_constraints() const;
+  int normalize_widen(int w) const;
+#ifdef ASSERT
+  bool contains(S v) const;
+  void verify_constraints() const;
+#endif // ASSERT
+};
+
+// Various helper functions for TypeInt/TypeLong operations
+class TypeIntHelper {
+public:
+  // Calculate the cardinality of a TypeInt/TypeLong ignoring the bits
+  // constraints, the return value is the cardinality minus 1 to not overflow
+  // with the bottom type
+  template <class S, class U>
+  static U cardinality_from_bounds(const RangeInt<S>& srange, const RangeInt<U>& urange) {
+    static_assert(std::is_signed<S>::value, "");
+    static_assert(std::is_unsigned<U>::value, "");
+    static_assert(sizeof(S) == sizeof(U), "");
+
+    if (U(srange._lo) == urange._lo) {
+      // srange is the same as urange
+      assert(U(srange._hi) == urange._hi, "");
+      return urange._hi - urange._lo;
+    }
+
+    // srange intersects with urange in 2 intervals [srange._lo, urange._hi]
+    // and [urange._lo, srange._hi]
+    return (urange._hi - U(srange._lo)) + (U(srange._hi) - urange._lo) + 1;
+  }
+
+  template <class CT, class S, class U>
+  static const Type* int_type_xmeet(const CT* i1, const Type* t2, const Type* (*make)(const TypeIntPrototype<S, U>&, int, bool), bool dual);
+
+  template <class CT>
+  static bool int_type_is_equal(const CT* t1, const CT* t2) {
+    return t1->_lo == t2->_lo && t1->_hi == t2->_hi &&
+           t1->_ulo == t2->_ulo && t1->_uhi == t2->_uhi &&
+           t1->_bits._zeros == t2->_bits._zeros && t1->_bits._ones == t2->_bits._ones;
+  }
+
+  template <class CT>
+  static bool int_type_is_subset(const CT* super, const CT* sub) {
+    return super->_lo <= sub->_lo && super->_hi >= sub->_hi &&
+           super->_ulo <= sub->_ulo && super->_uhi >= sub->_uhi &&
+           (super->_bits._zeros &~ sub->_bits._zeros) == 0 && (super->_bits._ones &~ sub->_bits._ones) == 0;
```

Comment:
Can you add a comment here what this does? Because it looks like a `&~` operator :)

I also wonder if it would make sense to delegate the `_bits` checks to its own class, for `equal` and `subset`.

Reply:
Yes it is indeed a `&~` operator because it is a set subtraction operation on the bitsets.

Reply:
> I also wonder if it would make sense to delegate the _bits checks to its own class, for equal and subset.

These are one liner and only used here so I don't think moving it away will improve readability.